src/frontend/bookmarks/collection.js

```diff
@@ -9,53 +9,33 @@ const getRawBookmarks = function() {
     });
 };
 
-const flattenBookmarks = function(rawBookmarks, list) {
-    return rawBookmarks.reduce(function(arr, treeNode) {
-        return arr.concat(treeNode.children ? flattenBookmarks(treeNode.children, []) : treeNode);
-    }, list);
-};
-
-const flattenTree = function(rawBookmarks, tree) {
-    return rawBookmarks.reduce(function(obj, treeNode) {
-        if (treeNode.children) {
-            obj[treeNode.id] = treeNode;
-            return flattenTree(treeNode.children, obj);
-        }
-        return obj;
-    }, tree);
-};
-
-const getPathUp = function(treeNode, map, path) {
-    const parent = map[treeNode.parentId];
-    if (parent.title && parent.id !== bookmarksBarId) {
-        path.unshift(parent.title);
-        return getPathUp(parent, map, path);
+const handlePath = function(treeNode, path) {
+    if (treeNode.title && treeNode.id !== bookmarksBarId) {
+        path.push(treeNode.title);
     }
-    return path;
 };
 
-const buildMap = function(map) {
-    return Object.keys(map).reduce(function(obj, key) {
-        const treeNode = map[key];
-        if (treeNode.parentId) {
-            obj[key] = getPathUp(treeNode, map, [treeNode.title]).join('/');
+const flattenBookmarks = function(rawBookmarks, list, path) {
+    const reduced = rawBookmarks.reduce(function(arr, treeNode) {
```

Comment:
![](http://cdn.meme.am/instances/58684173.jpg)
may be bookmarks instead of arr?